File: infrastructure/docker/Dockerfile.cpp
# Multi-stage build for C++ AVX-512 Engine
# Stage 1: Builder
FROM gcc:13.2 AS builder

WORKDIR /build

# Copy C++ source files
COPY backend/cpp_engine/*.cpp .
COPY backend/cpp_engine/*.hpp .

# Compile with aggressive optimizations
RUN g++ -O3 -std=c++23 -march=native -mavx512f -mavx512dq \
    -pthread -flto -fno-exceptions -fno-rtti \
    -o rct_core rct_core.cpp

# Stage 2: Minimal runtime
FROM alpine:3.23

RUN apk add --no-cache libstdc++

COPY --from=builder /build/rct_core /usr/local/bin/rct_core

EXPOSE 9000

CMD ["/usr/local/bin/rct_core"]

File: infrastructure/docker/Dockerfile.go
# Multi-stage build for Go Billing Exporter
FROM golang:1.22-alpine AS builder

WORKDIR /app

# Copy go mod files
COPY backend/go_exporter/go.mod backend/go_exporter/go.sum ./
RUN go mod download

# Copy source code
COPY backend/go_exporter/*.go ./

# Build binary with optimizations
RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build \
    -ldflags="-s -w -X main.version=$(git describe --tags)" \
    -o rct_exporter .

# Final stage
FROM alpine:3.24

RUN apk add --no-cache ca-certificates

COPY --from=builder /app/rct_exporter /usr/local/bin/rct_exporter

EXPOSE 9102

USER 65534:65534

ENTRYPOINT ["/usr/local/bin/rct_exporter"]
